### PR TITLE
Task: tags for events

### PR DESCRIPTION
**Pull request by DawidekZagajnik**

### CODE REVIEW

File: tracardi/service/storage/drivers/elastic/tag.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+from typing import List
+
+
+async def get_document_by_event_type(event_type):
+    return await storage_manager("event-tags").load_by(field="type", value=event_type)
+
+
+async def add_tags(event_type: str, tags: List[str]):
+    search_result = await get_document_by_event_type(event_type=event_type)
+    if search_result.dict()["result"]:
```
Comment (atompie) [inline tracardi/service/storage/drivers/elastic/tag.py:11]: move search_result.dict() to variable. Making dict() form pydantic object can be expensive.

File: tracardi/service/storage/drivers/elastic/tag.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        tags.extend(search_result.dict()["result"][0]["tags"])
+        tags = list(set(tags))
+        return 1, await storage_manager("event-tags").upsert({
+            "_id": search_result.dict()["result"][0]["id"],
```
Comment (atompie) [inline tracardi/service/storage/drivers/elastic/tag.py:15]: search_result.dict() is repeated multiple times.

File: tracardi/service/storage/drivers/elastic/tag.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+            "type": search_result.dict()["result"][0]["type"],
+            "tags": tags
+        })
+    else:
+        return 0, await storage_manager("event-tags").upsert({
```
Comment (atompie) [inline tracardi/service/storage/drivers/elastic/tag.py:20]: What 0 and 1 means in return. Maybe you could add some documentations so users can understand it?

File: tracardi/service/storage/drivers/elastic/tag.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                await storage_manager("event-tags").upsert({
+                    "_id": search_result.dict()["result"][0]["id"],
+                    "type": event_type,
+                    "tags": tags_to_insert
+                })
+    else:
+        raise ValueError("There is no document with 'type' field equal to '{}'.".format(event_type))
+
+
+async def load_all_tags(limit):
```
Comment (atompie) [inline tracardi/service/storage/drivers/elastic/tag.py:46]: Add default value to limit

File: tracardi/service/storage/drivers/elastic/tag.py (hunk 3 above)
Comment (atompie) [reply tracardi/service/storage/drivers/elastic/tag.py:20]: I saw in tracardi API that this is the number of updated records. What happens if there is an error. Elastic returns the number of updated records.

**Review by atompie — CHANGES_REQUESTED** [review]

File: tracardi/service/storage/drivers/elastic/tag.py [hunk 5]
```diff
@@ -8,40 +8,42 @@ async def get_document_by_event_type(event_type):
 
 async def add_tags(event_type: str, tags: List[str]):
```
Comment (atompie) [inline tracardi/service/storage/drivers/elastic/tag.py:9]: See the change that is required in API. Please simplify the names of functions.

File: tracardi/service/storage/drivers/elastic/tag.py [hunk 6]
```diff
@@ -8,40 +8,42 @@ async def get_document_by_event_type(event_type):
...
     search_result = await get_document_by_event_type(event_type=event_type)
-    if search_result.dict()["result"]:
-        tags.extend(search_result.dict()["result"][0]["tags"])
+    result_dict = search_result.dict()
```
Comment (atompie) [inline tracardi/service/storage/drivers/elastic/tag.py:11]: I do not know why you love dicts. Dicts are the worse possible solution for managing data. That is why driver is returning objects. If you are lacking some functionality in object then you can modify it. But most of the time you do not need that.

See How you can use object to do the same you did:

```
async def add_tags(event_type: str, tags: List[str]):
    search_result = await get_document_by_event_type(event_type=event_type)
    if search_result.total == 1:  # Now we are sure that there is only on record per event type
        record = list(search_result).pop()
        # TagForm should be moved from API to tracardi and named EventTag
        tag = EventTag(**record)  # now record is validated
        tag.tags.extend(tags)  # Appended tags
        tag.tags = list(set(tag.tags))  # Make it uniq
        return await storage_manager("event-tags").upsert(tag.dict())  # use object to save data
    else:
        return await storage_manager("event-tags").upsert({
            "type": event_type,
            "tags": tags
        })
```

Comment (atompie) [reply tracardi/service/storage/drivers/elastic/tag.py:11]: It is a lot of cleaner and is not so error prone as dicts.

Comment (atompie) [reply tracardi/service/storage/drivers/elastic/tag.py:11]: You should not convert SearchObjects to dict to process it.

**Review by atompie — CHANGES_REQUESTED** [review]

File: tracardi/service/storage/drivers/elastic/tag.py (hunk 6 above)
Comment (DawidekZagajnik) [reply tracardi/service/storage/drivers/elastic/tag.py:11]: where exactly should TagForm be moved to? (I mean directory or file name)

Comment (DawidekZagajnik) [reply tracardi/service/storage/drivers/elastic/tag.py:11]: never mind, I've got it in second PR

Comment (atompie) [reply tracardi/service/storage/drivers/elastic/tag.py:11]: > where exactly should TagForm be moved to? (I mean directory or file name)

To tracardi repo, folder tracardi/domain

**Comment by DawidekZagajnik** [comment]
please ignore changes in event mappings for now, that's for next issues, in same repo though since that's one topic (tags)

File: tracardi/service/tags_service.py [hunk 7]
```diff
@@ -0,0 +1,4 @@
+from tracardi.event_server.utils.memory_cache import MemoryCache
```
Comment (atompie) [inline tracardi/service/tags_service.py:1]: Do not know why this is implemented?

**Review by atompie — CHANGES_REQUESTED** [review]

File: tracardi/service/tags_service.py (hunk 7 above)
Comment (DawidekZagajnik) [reply tracardi/service/tags_service.py:1]: gonna delete this

Comment (DawidekZagajnik) [reply tracardi/service/tags_service.py]: deleted, sorry for mistake

**Review by atompie — APPROVED** [review]

**Pull request merged**